components/cam/src/physics/rrtmgp/cam_optics.F90

```diff
@@ -724,53 +497,37 @@ subroutine set_aerosol_optics_sw(icall, state, pbuf, &
                             count(night_indices > 0), night_indices, is_cmip6_volc, &
                             tau, tau_w, tau_w_g, tau_w_f)
 
-      ! Reset outputs (also handles case where radiation grid contains an extra
-      ! layer above CAM grid)
-      optics_out%tau = 0
-      optics_out%ssa = 1
-      optics_out%g = 0
-
-      ! Assign daytime columns
-      do iday = 1,count(day_indices > 0)
-
-         ! Get index into full chunk-wide array for this daytime index
-         icol = day_indices(iday)
-
+      ! Extract quantities from products
+      do icol = 1,ncol
          ! Copy cloud optical depth over directly
-         optics_out%tau(iday,ktop:kbot,1:nswbands) = tau(icol,1:pver,1:nswbands)
-
+         tau_out(icol,1:pver,1:nswbands) = tau(icol,1:pver,1:nswbands)
          ! Extract single scattering albedo from the product-defined fields
          where (tau(icol,1:pver,1:nswbands) > 0)
-            optics_out%ssa(iday,ktop:kbot,1:nswbands) &
+            ssa_out(icol,1:pver,1:nswbands) &
                = tau_w(icol,1:pver,1:nswbands) / tau(icol,1:pver,1:nswbands)
          elsewhere
-            optics_out%ssa(iday,ktop:kbot,1:nswbands) = 1
+            ssa_out(icol,1:pver,1:nswbands) = 1
          endwhere
-
          ! Extract assymmetry parameter from the product-defined fields
          where (tau_w(icol,1:pver,1:nswbands) > 0)
-            optics_out%g(iday,ktop:kbot,1:nswbands) &
+            asm_out(icol,1:pver,1:nswbands) &
                = tau_w_g(icol,1:pver,1:nswbands) / tau_w(icol,1:pver,1:nswbands)
          elsewhere
-            optics_out%g(iday,ktop:kbot,1:nswbands) = 0
+            asm_out(icol,1:pver,1:nswbands) = 0
          endwhere
-
       end do
 
       ! We need to fix band ordering because the old input files assume RRTMG band
       ! ordering, but this has changed in RRTMGP.
       ! TODO: fix the input files themselves!
-      do icol = 1,size(optics_out%tau,1)
-         do ilay = 1,size(optics_out%tau,2)
-            optics_out%tau(icol,ilay,:) = reordered(optics_out%tau(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
-            optics_out%ssa(icol,ilay,:) = reordered(optics_out%ssa(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
-            optics_out%g(icol,ilay,:) = reordered(optics_out%g(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
+      do icol = 1,size(tau_out,1)
+         do ilay = 1,size(tau_out,2)
+            tau_out(icol,ilay,:) = reordered(tau_out(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
+            ssa_out(icol,ilay,:) = reordered(ssa_out(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
+            asm_out(icol,ilay,:) = reordered(asm_out(icol,ilay,:), map_rrtmg_to_rrtmgp_swbands)
          end do
       end do
 
-      ! Check values
-      call handle_error(optics_out%validate())
```

Comment:
Should we have a value check here?

Reply:
It gets checked later anyways, this existed for debugging.